src/AWS.Deploy.ServerMode.Client/ServerModeSession.cs

```diff
@@ -0,0 +1,122 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.\r
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using AWS.Deploy.ServerMode.Client.Exceptions;
+
+namespace AWS.Deploy.ServerMode.Client
+{
+    /// <summary>
+    /// Helper class that allows launching deployment tool in server mode.
+    /// </summary>
+    public class ServerModeSession
+    {
+        private const int TCP_PORT_ERROR = -100;
+
+        private readonly int _startPort;
+        private readonly int _endPort;
+
+        public ServerModeSession(int startPort = 5000, int endPort = 5100)
+        {
+            _startPort = startPort;
+            _endPort = endPort;
+        }
+
+        /// <summary>
+        /// Start deployment tool in server mode.
+        /// </summary>
+        /// <returns>Base URL of the  deployment tool local server.</returns>
+        /// <exception cref="InternalServerModeException">Throws when deployment tool server failed to start for un unknown reason.</exception>
+        /// <exception cref="PortUnavailableException">Throws when deployment tool server failed to start due to unavailability of free ports.</exception>
+        public async Task<string> Start()
+        {
+            var currentProcessId = Process.GetCurrentProcess().Id;
+
+            for (var port = _startPort; port <= _endPort; port++)
+            {
+                var baseUrl = $"http://localhost:{port}";
+
+                var command = $"dotnet aws server-mode --port {port} --parent-pid {currentProcessId}";
+                var startServerTask = RunCommand(command);
+
+                var isServerAvailableTask = IsServerAvailable(baseUrl);
+
+                // If the server is server health check succeeded
+                // return the baseUrl
+                if (isServerAvailableTask == await Task.WhenAny(startServerTask, isServerAvailableTask))
+                {
+                    return baseUrl;
+                }
+
+                // For -100 errors, we want to check all the ports in the configured port range
+                // If the error code other than -100, this is an unexpected exit code.
+                if (startServerTask.Result != TCP_PORT_ERROR)
+                {
+                    throw new InternalServerModeException($"\"{command}\" failed for unknown reason.");
```

Comment:
nit: I'd rather we not specify the command explicitly, just provide a more descriptive message.

Reply:
This is for VS Toolkit, I don't see a reason not to specify the command. It will never be shown to the customer. Also, we dont't know in this case, what is the underlying reason, therefore, having command atleast make debugging easier.